python/ray/train/_internal/checkpoint.py

```diff
@@ -102,14 +102,27 @@ def _process_checkpoint(
         """Ray Train entrypoint. Perform all processing for a checkpoint."""
         # Get checkpoint from first worker.
         checkpoint_data = checkpoint_results[0].data
-
-        # Decode checkpoint.
-        checkpoint_data = decode_checkpoint_fn(checkpoint_data)
+        checkpoint_metadata = checkpoint_results[0].metadata or {}
+        encoded = checkpoint_results[0].encoded
+
+        # Decode checkpoint if it's not persisted
+        if encoded:
+            checkpoint_data = decode_checkpoint_fn(checkpoint_data.to_dict())
```

Comment:
I think we can always call `decode_checkpoint_fn` regardless of if the checkpoint was encoded.

You can imagine the Backend has decode overridden but not encode.